modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -621,174 +299,110 @@ public CompletableFuture<Set<String>> dataNodes(long causalityToken, int zoneId)
-                // zone creation, already handled in a separate listener.
-                return completedFuture(null);
-            }
-
-            DistributionZoneView zoneView = ctx.newValue(DistributionZoneView.class);
+    private CompletableFuture<Void> onUpdateFilter(AlterZoneEventParameters parameters) {
+        int zoneId = parameters.zoneDescriptor().id();
 
-            int zoneId = zoneView.zoneId();
+        String newFilter = parameters.zoneDescriptor().filter();
 
-            String filter = zoneView.filter();
+        long causalityToken = parameters.causalityToken();
 
-            long revision = ctx.storageRevision();
+        VaultEntry filterUpdateRevision = vaultMgr.get(zonesFilterUpdateRevision()).join();
 
-            VaultEntry filterUpdateRevision = vaultMgr.get(zonesFilterUpdateRevision()).join();
-
-            if (filterUpdateRevision != null) {
-                // This means that we have already handled event with this revision.
-                // It is possible when node was restarted after this listener completed,
-                // but applied revision didn't have time to be propagated to the Vault.
-                if (bytesToLong(filterUpdateRevision.value()) >= revision) {
-                    return completedFuture(null);
-                }
+        if (filterUpdateRevision != null) {
+            // This means that we have already handled event with this causalityToken.
+            // It is possible when node was restarted after this listener completed,
+            // but applied causalityToken didn't have time to be propagated to the Vault.
+            if (bytesToLong(filterUpdateRevision.value()) >= causalityToken) {
+                return completedFuture(null);
             }
-
-            vaultMgr.put(zonesFilterUpdateRevision(), longToBytes(revision)).join();
-
-            saveDataNodesToMetaStorageOnScaleUp(zoneId, revision);
-
-            causalityDataNodesEngine.onUpdateFilter(revision, zoneId, filter);
-
-            return completedFuture(null);
-        };
-    }
-
-    private class ZonesConfigurationListener implements ConfigurationNamedListListener<DistributionZoneView> {
-        @Override
-        public CompletableFuture<?> onCreate(ConfigurationNotificationEvent<DistributionZoneView> ctx) {
-            DistributionZoneView zone = ctx.newValue();
-
-            createOrRestoreZoneState(zone, ctx.storageRevision());
-
-            return completedFuture(null);
         }
 
-        @Override
-        public CompletableFuture<?> onDelete(ConfigurationNotificationEvent<DistributionZoneView> ctx) {
-            int zoneId = ctx.oldValue().zoneId();
-
-            long revision = ctx.storageRevision();
-
-            ZoneState zoneState = zonesState.get(zoneId);
+        vaultMgr.put(zonesFilterUpdateRevision(), longToBytes(causalityToken)).join();
 
-            zoneState.stopTimers();
+        saveDataNodesToMetaStorageOnScaleUp(zoneId, causalityToken);
```

Comment:
Well, it's not yours however in order to prevent ms invocations reordering we should return saveDataNodesToMetaStorageOnScaleUp result future as onAnyCallback result.

Reply:
fix it